games/xworld3d/tasks/XWorld3DNavTargetSide.py

```diff
@@ -27,57 +28,141 @@ def __init__(self, env):
+        empty_grids = set(empty_grids1 + empty_grids2 + [agent.loc])
+        valid_empty_grids = [e for e in empty_grids \
+                             if self._reachable(agent.loc, e) \
+                             and self.__compute_triple_direction(pair[0], pair[1], e)]
+
+        assert valid_empty_grids, "map too crowded?"
+        direction = self.__compute_triple_direction(
+            pair[0], pair[1], random.choice(valid_empty_grids))
+        self._record_target((pair[0], pair[1], direction));
+        self._bind("S -> start")
+        self._bind("G1 -> '" + pair[0].name + "'")
+        self._bind("G2 -> '" + pair[1].name + "'")
+        self._bind("P -> " + direction.upper())
+        self.sentence = self._generate()
+        return ["navigation_reward", 0.0, self.sentence]
+
+    def navigation_reward(self):
+        threshold = 1.5
+        reward, time_out = self._time_reward()
+        if not time_out:
+            agent, _, _ = self._get_agent()
+            objects_reach_test = [g.id for g in self._get_goals() \
+                                  if self._reach_object(agent.loc, agent.yaw, g)]
+            if objects_reach_test:
+                reward = self._failed_goal(reward)
+            else:
+                g1, g2, target_direction = self.target
+                direction = self.__compute_triple_direction(g1, g2, agent.loc, agent.yaw)
+                if direction == target_direction \
+                   and min(self._get_direction_and_distance(agent.loc, g1.loc),
+                           self._get_direction_and_distance(agent.loc, g2.loc)) < threshold:
+                    reward = self._successful_goal(reward)
+
+        return ["navigation_reward", reward, self.sentence]
+
+    def __compute_triple_direction(self, g1, g2, a, view_yaw=None):
+        """
+        Compute the direction of g1 to g2 in the agent's view
+        """
+        mid = self._middle_loc(g1.loc, g2.loc)
+        if view_yaw is None:
+            view_yaw = math.atan2(mid[1] - a[1], mid[0] - a[0])
+
+        # agent must look at the two objects
+        if abs(self._get_direction_and_distance(
+                a, mid, view_yaw)[0]) > XWorld3DTask.PI_8:
+            return False
+
+        theta, _, _ = self._get_direction_and_distance(g1.loc, g2.loc, view_yaw)
+        sign = True if theta > 0 else False
+        flag = False # g2 is far
+        theta = abs(theta)
+        if theta > XWorld3DTask.PI_2:
+            flag = True # g1 is far
+            theta = XWorld3DTask.PI - theta
+
+        ##
+        v1 = (1, 1)
+        v2 = (0.5, 1.5)
+        sep_theta = math.acos((v1[0]*v2[0] + v1[1]*v2[1]) \
+                              / math.sqrt(2) / math.sqrt(0.25 + 2.25))
+
+        if theta < sep_theta + 1e-3: # front or behind
+            if flag:
+                return "front"
+            else:
+                return "behind"
+#        elif XWorld3DTask.PI_2 - theta < XWorld3DTask.PI_4 + 1e-3: # left or right
```

Comment:
remove

Reply:
Not for now. Might uncomment in the future.